### ISSUE TITLE
Implement Gaussian Process as a model

### ISSUE THREAD

**Issue by oesteban** [nipreps/eddymotion#53]
In order to compare baseline performance vs. FSL eddy.

**Comment by josephmje** [comment]
Adding notes from our 2021-12-07 meeting here:

A first stab at implementing the Gaussian process model is at #60. 

At the moment, we haven't done any tuning or optimization of the model. If we are going to eventually compare it to FSL eddy, that will need to be done.

For testing, some suggestions are to use:
- simulated data from FiberFox
- [IXI dataset](https://osf.io/8k95s)
- a high quality dataset with certain voxels masked (perhaps near a white matter tract or some CSF in the middle of a ventricle)

Will eventually need to add better parallelization. Currently, the DTI model chunks the data before fitting the model. We can also implement that code here. For the future, we may need to dig into dipy and can use joblib/dask to further improve the parallelization.

**Comment by dPys** [comment]
@josephmje - Not sure if you or anyone else are still working on the GP model, but figured I'd chime in with a few thoughts to get the conversation going again:

The GP model needs at least a PoC design for the kernel before it can be tuned. Here's a first template for what that might look like. High-Level Requirements:
*Isotropism, uniform affinity for all angular directions.
*Symmetry around the origin.

The way that most folks have seemed to address this in the context of q-space is to factorize the covariance function into (a) an angular part and (b) a radial part, with the addition of (c) an error term. 

The angular part can be constructed using Legendre polynomials, whereby symmetry about the origin can be achieved by excluding odd terms and ensuring that their coefficients are positive. Past work seems to have used an order <=6, with the consensus being that grabbing the first four terms is best.

Maybe roughly something along the lines of:
```
from numpy.polynomial.legendre import Legendre

def angular_polynomials(n_gradients, order=6):
    """
    A real continuous function $C(\theta)$ is a valid covariance function on the 
    sphere if and only if it is of the form: 
    $\sum_{n=0}^{\infty} a_{n}P_{n}(\cos(\theta))$ 
    where $\theta \in (0, \pi)$
    """
    theta = np.arange(n_gradients)/n_gradients * np.pi
    num_pol = range(0, order + 1, 2)
    y = np.ones((len(num_pol),len(theta)))   
    coeff = np.eye(int(order/2 + 1))
    
    for i, n in enumerate(num_pol):
        leg = Legendre(np.abs(coeff[i]))
        y[i] = leg(np.cos(theta))
    return y
```

For the radial part, we probably wouldn't be able to get away with sklearn's standard RBF kernel, but we might look to https://github.com/dipy/dipy/blob/a35b72d8d86c3426922cee97063e35184c3785bb/dipy/core/interpolation.pyx#L14-L85 for inspiration?

and then the final kernel would look something like the following, with ~ six hyperparameters that'd need to be tuned:
```
kernel =  RBF(length_scale=ls, length_scale_bounds=(ls_l, ls_u)) * angular_polynomials(np.sum(~gtab.b0s_mask)) + kernels.WhiteKernel(noise_level=nl, noise_level_bounds=(nl_l, nl_u))
```

The bulk of the work on this would be the initial round of tuning done during a pre-training stage, and using a synthetic, fiber-fox generated sample as you originally had proposed. The objective that we'd want to minimize is the negative log marginal likelihood, and the data itself would be synthesized such that training and validation sets have similar amounts of noise / order of magnitude but perhaps different numbers of directions compared to evaluation data.

A ton of work is involved in this obviously, but figured it couldn't hurt to discuss more. WDYT?

**Comment by oesteban** [comment]
A base implementation of the GP using sklearn is now in the codebase. This has been addressed :tada:
